### ISSUE TITLE
lock account after X bad passwords

### ISSUE THREAD

**Issue by benadida** [mozilla/persona#2656]
After X bad passwords, the Persona account should be locked. Simplest agreed-upon behavior: after 3 bad passwords, treat account as unverified (e.g. password reset).

We might want to add a CAPTCHA to prevent denial-of-service, but in that case we want to add it to "forgot password," too, so that might not be as urgent.

**Comment by zaach** [comment]
We discussed using a router monitor for handling brute force attacks, which is dependent on #2508.

**Comment by seanmonstar** [comment]
Way too busy on bigtent and b2g, removing myself as assignee

**Comment by ygjb** [comment]
Can we get an update on when this feature will be implemented?

**Comment by benadida** [comment]
@lloyd @zaach can we prioritize this issue in the next 3-4 weeks?

**Comment by benadida** [comment]
alright, I'm going to see if I can provide some assistance on this one. I'm thinking the router approach may not be the right one, as that would mean adding smarts in the router about which URLs do what... which seems like the wrong direction. @zaach please shout if you think I'm missing something here.

**Comment by shane-tomlinson** [comment]
@benadida - is this still on the priority queue?

**Comment by benadida** [comment]
Yes, very much so.

**Comment by lloyd** [comment]
@fmarier yvan wants this so bad he can taste it....

**Comment by fmarier** [comment]
@lloyd and I discussed this and we propose keeping this as simple as possible:
- new `failedpasswords` column in the user table
- that column is incremented by one every time a user submits the wrong password
- if that count reaches 10, then we clear the password in the DB and reset the failed count to 0
- next time the user tries to login, they get the "forgot password" flow and my do an email verification & set a new password

**Comment by shane-tomlinson** [comment]
@fmarier - I am hesitant to wipe the user's password, this could lead alice, as somebody that doesn't like bob, to cause bob to have to reset his password over and over again.

What about when the password count reaches 10, lock the user out for several minutes and send an email with a link to optionally reset the password?

**Comment by lloyd** [comment]
after spending time with the code, I'd like to take a slight tweak on @fmarier proposal:
- if that count reaches 10, then when they try to authenticate they see "your account is locked, please click forgot password to reset your password"

this is identical in spirit but something we can execute quickly.

@shane-tomlinson in the future we can add captcha or some nicer option, given the severity of this and how much the security team wants it, I think it is better to allow alice to be a jerk for the short term.

**Comment by lloyd** [comment]
remaining work:
1. dbwriter api call
2. add configuration parameter
3. update authenticate_user api to check failed password attempts and return a specific error
4. update frontend to display a different error tooltip based on server response in 3

now that database changes and tests are implemented, this is about another hour or two of dev work.

**Comment by lloyd** [comment]
done.  r? ^^

**Comment by lloyd** [comment]
merged into dev.  to be included in http://personatra.in/#2013.05.10

**Comment by jrgm** [comment]
verified train-2013.05.22

**Comment by lloyd** [comment]
woohoo
